hail/src/main/scala/is/hail/expr/ir/MatrixWriter.scala

```diff
@@ -39,6 +61,255 @@ case class MatrixNativeWriter(
+  def ifIndexed[T >: Null](obj: => T): T = if (hasIndex) obj else null
+
+  def consumeStream(
+    context: EmitCode,
+    eltType: PStruct,
+    mb: EmitMethodBuilder[_],
+    region: Value[Region],
+    stream: SizedStream): EmitCode = {
+    val enc1 = spec1.buildEmitEncoderF(eltType, mb.ecb, typeInfo[Long]) //(Value[Region], Value[T], Value[OutputBuffer]) => Code[Unit]
+    val enc2 = spec2.buildEmitEncoderF(eltType, mb.ecb, typeInfo[Long]) //(Value[Region], Value[T], Value[OutputBuffer]) => Code[Unit]
+
+    val keyType = ifIndexed { index.get._2 }
+    val iAnnotationType = +PCanonicalStruct("entries_offset" -> +PInt64())
+    val indexWriter = ifIndexed { StagedIndexWriter.withDefaults(keyType, mb.ecb, annotationType = iAnnotationType) }
+
+    context.map { ctxCode: PCode =>
+      val ctx = mb.newLocal[Long]("ctx")
+      val result = mb.newLocal[Long]("write_result")
+
+      val filename1 = mb.newLocal[String]("filename1")
+      val os1 = mb.newLocal[ByteTrackingOutputStream]("write_os1")
+      val ob1 = mb.newLocal[OutputBuffer]("write_ob1")
+      val filename2 = mb.newLocal[String]("filename2")
+      val os2 = mb.newLocal[ByteTrackingOutputStream]("write_os2")
+      val ob2 = mb.newLocal[OutputBuffer]("write_ob2")
+      val n = mb.newLocal[Long]("partition_count")
+
+      def writeFile(codeRow: EmitCode): Code[Unit] = {
+        val rowType = coerce[PStruct](codeRow.pt)
+        EmitCodeBuilder.scopedVoid(mb) { cb =>
+          codeRow.toI(cb).consume(cb,
+            cb._fatal("row can't be missing"),
+            { pc =>
+              val row = pc.memoize(cb, "row")
+              if (hasIndex) {
+                val keyRVB = new StagedRegionValueBuilder(mb, keyType)
+                val aRVB = new StagedRegionValueBuilder(mb, iAnnotationType)
+                indexWriter.add(cb, {
+                  cb += keyRVB.start()
+                  keyType.fields.foreach { f =>
+                    cb += keyRVB.addIRIntermediate(f.typ)(Region.loadIRIntermediate(f.typ)(rowType.fieldOffset(coerce[Long](row.value), f.name)))
+                    cb += keyRVB.advance()
+                  }
+                  IEmitCode.present(cb, PCode(keyType, keyRVB.offset))
+                }, ob1.invoke[Long]("indexOffset"), {
+                  cb += aRVB.start()
+                  cb += aRVB.addLong(ob2.invoke[Long]("indexOffset"))
+                  IEmitCode.present(cb, PCode(iAnnotationType, aRVB.offset))
+                })
+              }
+              cb += ob1.writeByte(1.asInstanceOf[Byte])
+              cb += enc1(region, coerce[Long](row.value), ob1)
+              cb += ob2.writeByte(1.asInstanceOf[Byte])
+              cb += enc2(region, coerce[Long](row.value), ob2)
+              cb.assign(n, n + 1L)
+            })
+        }
+      }
+
+      PCode(pResultType, EmitCodeBuilder.scopedCode(mb) { cb: EmitCodeBuilder =>
+        cb.assign(ctx, ctxCode.tcode[Long])
```

Comment:
I've fixed the ctx variable. For result, os1 and os2, I don't think I see a way to memoize them, because they're not actually PCodes---is there a way to memoize arbitrary Java objects?

Reply:
Ah, you're right. This is fine.